File: worker/src/sessionLease.ts
import type { SupabaseClient } from '@supabase/supabase-js'
import { listenerWorkerId, leaseRoleLabel, workerConfig } from './workerConfig'

const LEASE_TTL_MS = Math.max(
  15_000,
  Math.min(120_000, Number(process.env.WORKER_SESSION_LEASE_TTL_MS ?? 45_000)),
)

export interface SessionLeaseRow {
  user_id: string
  worker_id: string
  role: string
  shard_id: number
  shard_count: number
  expires_at: string
}

function expiresAtIso(): string {
  return new Date(Date.now() + LEASE_TTL_MS).toISOString()
}

/**
 * Claim listener ownership for user_id. Fails if another worker holds a non-expired lease.
 */
export async function acquireSessionLease(
  supabase: SupabaseClient,
  userId: string,
): Promise<{ ok: true } | { ok: false; reason: string }> {
  const workerId = listenerWorkerId()
  const now = new Date().toISOString()

  const { data: existing } = await supabase
    .from('worker_session_leases')
    .select('worker_id, expires_at')
    .eq('user_id', userId)
    .maybeSingle()

  if (existing) {
    const exp = new Date(existing.expires_at as string).getTime()
    const held = existing.worker_id as string
    if (exp > Date.now() && held !== workerId) {
      return { ok: false, reason: `lease held by ${held} until ${existing.expires_at}` }
    }
  }

  const { error } = await supabase.from('worker_session_leases').upsert(
    {
      user_id: userId,
      worker_id: workerId,
      role: leaseRoleLabel(),
      shard_id: workerConfig.shardId,
      shard_count: workerConfig.shardCount,
      expires_at: expiresAtIso(),
      updated_at: now,
    },
    { onConflict: 'user_id' },
  )

  if (error) return { ok: false, reason: error.message }
  return { ok: true }
}

export async function renewSessionLease(supabase: SupabaseClient, userId: string): Promise<void> {
  const workerId = listenerWorkerId()
  await supabase
    .from('worker_session_leases')
    .update({
      worker_id: workerId,
      expires_at: expiresAtIso(),
      updated_at: new Date().toISOString(),
    })
    .eq('user_id', userId)
    .eq('worker_id', workerId)
}

export async function releaseSessionLease(supabase: SupabaseClient, userId: string): Promise<void> {
  const workerId = listenerWorkerId()
  await supabase
    .from('worker_session_leases')
    .delete()
    .eq('user_id', userId)
    .eq('worker_id', workerId)
}

/** Trade workers: true when a listener shard holds a fresh lease (Telegram path is live). */
export async function isTelegramListenerLiveForUser(
  supabase: SupabaseClient,
  userId: string,
): Promise<boolean> {
  const { data } = await supabase
    .from('worker_session_leases')
    .select('expires_at, role')
    .eq('user_id', userId)
    .maybeSingle()

  if (!data) return false
  const role = String(data.role ?? '')
  if (role !== 'listener' && role !== 'all') return false
  return new Date(data.expires_at as string).getTime() > Date.now()
}

export async function listActiveLeases(
  supabase: SupabaseClient,
): Promise<SessionLeaseRow[]> {
  const { data } = await supabase
    .from('worker_session_leases')
    .select('*')
    .gt('expires_at', new Date().toISOString())
  return (data ?? []) as SessionLeaseRow[]
}

File: telegram-listener/app/session_lease.py
"""Postgres session lease helpers."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any

from supabase import Client

from .config import Config, lease_role_label, listener_worker_id


def _expires_at_iso(cfg: Config) -> str:
    return (datetime.now(timezone.utc) + timedelta(milliseconds=cfg.lease_ttl_ms)).isoformat()


async def acquire_session_lease(supabase: Client, cfg: Config, user_id: str) -> tuple[bool, str]:
    worker_id = listener_worker_id(cfg)
    expires_at = _expires_at_iso(cfg)

    try:
        result = supabase.rpc(
            "acquire_worker_session_lease",
            {
                "p_user_id": user_id,
                "p_worker_id": worker_id,
                "p_role": lease_role_label(cfg),
                "p_shard_id": cfg.shard_id,
                "p_shard_count": cfg.shard_count,
                "p_expires_at": expires_at,
            },
        ).execute()
        if result.data is True:
            return True, ""
    except Exception as exc:
        return _acquire_legacy(supabase, cfg, user_id, str(exc))

    existing = (
        supabase.table("worker_session_leases")
        .select("worker_id, expires_at")
        .eq("user_id", user_id)
        .maybe_single()
        .execute()
    )
    row = existing.data
    if row:
        held = row.get("worker_id", "")
        exp = row.get("expires_at", "")
        return False, f"lease held by {held} until {exp}"
    return False, "lease acquire rejected"


def _acquire_legacy(supabase: Client, cfg: Config, user_id: str, rpc_err: str) -> tuple[bool, str]:
    worker_id = listener_worker_id(cfg)
    existing = (
        supabase.table("worker_session_leases")
        .select("worker_id, expires_at")
        .eq("user_id", user_id)
        .maybe_single()
        .execute()
    )
    row = existing.data
    if row:
        exp = datetime.fromisoformat(str(row["expires_at"]).replace("Z", "+00:00"))
        held = str(row.get("worker_id", ""))
        if exp.timestamp() > datetime.now(timezone.utc).timestamp() and held != worker_id:
            return False, f"lease held by {held} until {row.get('expires_at')}"

    supabase.table("worker_session_leases").upsert(
        {
            "user_id": user_id,
            "worker_id": worker_id,
            "role": lease_role_label(cfg),
            "shard_id": cfg.shard_id,
            "shard_count": cfg.shard_count,
            "expires_at": _expires_at_iso(cfg),
            "updated_at": datetime.now(timezone.utc).isoformat(),
        },
        on_conflict="user_id",
    ).execute()
    return True, ""


def renew_session_lease(supabase: Client, cfg: Config, user_id: str) -> None:
    worker_id = listener_worker_id(cfg)
    supabase.table("worker_session_leases").update(
        {
            "worker_id": worker_id,
            "expires_at": _expires_at_iso(cfg),
            "updated_at": datetime.now(timezone.utc).isoformat(),
        }
    ).eq("user_id", user_id).eq("worker_id", worker_id).execute()


def release_session_lease(supabase: Client, cfg: Config, user_id: str) -> None:
    worker_id = listener_worker_id(cfg)
    supabase.table("worker_session_leases").delete().eq("user_id", user_id).eq(
        "worker_id", worker_id
    ).execute()


def list_active_leases(supabase: Client) -> list[dict[str, Any]]:
    now = datetime.now(timezone.utc).isoformat()
    result = supabase.table("worker_session_leases").select("*").gt("expires_at", now).execute()
    return list(result.data or [])
